core/sql/generator/GenRelUpdate.cpp

```diff
@@ -2844,8 +2848,8 @@ short HbaseInsert::codeGen(Generator *generator)
 	  (noCheck()))
 	hbasescan_tdb->setHbaseSqlIUD(FALSE);
 
-      if (((getInsertType() == Insert::VSBB_INSERT_USER) && 
-                   generator->oltOptInfo()->multipleRowsReturned()) ||
+      if (((((getInsertType() == Insert::VSBB_INSERT_USER) || isUpsert() )&& 
+           generator->oltOptInfo()->multipleRowsReturned())) ||
```

Comment:
A nit, there is an unnecessary pair of parentheses at the beginning of line 2851 and the end of line 2852.